sim-lib/src/lib.rs

```diff
@@ -826,41 +888,53 @@ async fn produce_events<N: DestinationGenerator + ?Sized, A: PaymentGenerator +
                     },
                     Err(e) => {
                         log::error!("Could not get amount for {source} -> {destination}: {e}. Please report a bug!");
-                        break;
+                        return Err(SimulationError::PaymentGenerationError(e));
                     },
                 };
 
                 log::debug!("Generated random payment: {source} -> {}: {amount} msat.", destination);
 
                 // Send the payment, exiting if we can no longer send to the consumer.
                 let event = SimulationEvent::SendPayment(destination.clone(), amount);
-                if let Err(e) = sender.send(event).await {
+                if let Err(e) = sender.send(event.clone()).await {
                     log::debug!(
                         "Stopped random producer for {amount}: {source} -> {destination}. Consumer error: {e}.",
                     );
-                    break;
+                    return Err(SimulationError::MpscChannelError { msg: format!("Error sending simulation event {event:?} to consumer"), error_str: e.to_string() });
                 }
+
             },
         }
     }
 
     log::debug!("Stopped random activity producer {source}.");
-    shutdown.trigger();
+    
+    Ok(())
 }
 
 async fn consume_simulation_results(
     logger: Arc<Mutex<PaymentResultLogger>>,
     receiver: Receiver<(Payment, PaymentResult)>,
     listener: Listener,
     write_results: Option<WriteResults>,
-) {
+) -> Result<(), SimulationError> {
     log::debug!("Simulation results consumer started.");
 
-    if let Err(e) = write_payment_results(logger, receiver, listener, write_results).await {
-        log::error!("Error while reporting payment results: {:?}.", e);
+    select! {
+        biased;
+        _ = listener.clone() => {
```

Comment:
Do we need to add a listener at this level if we've already got one in `write_payment_results`? 

Looking at this wish fresh eyes, not sure why we have two levels of functions here - can't we just `spawn consume_simulation_results` and move the code for `write_payment_results` into this function?

Reply:
`write_payment_results` has been moved to `consume_simulation_results`. Marking as resolved but reopen if you have any reservations.